lib/http_v4_signature.c

```diff
@@ -0,0 +1,325 @@
+#include "sendf.h"
+
+#include <time.h>
+
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+
+#define HMAC_SHA256(k, kl, d, dl, o)                                    \
+  if(Curl_hmacit(Curl_HMAC_SHA256, k, kl, d, dl, o) != CURLE_OK) {      \
+    ret = CURLE_OUT_OF_MEMORY;                                          \
+    goto free_all;                                                      \
+  }
+
+#define PROVIDER_MAX_L 16
+#define REQUEST_TYPE_L (PROVIDER_MAX_L + sizeof("4_request"))
+/* secret key is 40 bytes long + PROVIDER_MAX_L + \0 */
+#define FULL_SK_L (PROVIDER_MAX_L + 40 + 1)
+
+static void sha256_to_str(char *dst, unsigned char *sha)
+{
+  int i;
+
+  for(i = 0; i < 32; ++i) {
+    curl_msnprintf(dst + (i * 2), 3, "%02x", sha[i]);
+  }
+}
+
+CURLcode Curl_output_v4_signature(struct connectdata *conn, bool proxy)
+{
+  CURLcode ret = CURLE_OK;
+  char sk[FULL_SK_L] = {0};
+  struct Curl_easy *data = conn->data;
+  const char *customrequest = data->set.str[STRING_CUSTOMREQUEST];
+  const char *hostname = data->state.up.hostname;
+  struct tm info;
+  time_t rawtime;
+  /* aws is the default because some provider that are not amazone still use
+   * aws:amz as prefix
+   */
+  const char *provider = data->set.str[STRING_V4_SIGNATURE] ?
+    data->set.str[STRING_V4_SIGNATURE] : "aws:amz";
+  size_t provider_l = strlen(provider);
+  char low_provider0[PROVIDER_MAX_L + 1] = {0};
+  char low_provider[PROVIDER_MAX_L + 1] = {0};
+  char up_provider[PROVIDER_MAX_L + 1] = {0};
+  char mid_provider[PROVIDER_MAX_L + 1] = {0};
+  char *region = NULL;
+  char *uri = NULL;
+  char *api_type = NULL;
+  char date_iso[17];
+  char date[9];
+  char date_str[64];
+  const char *post_data = data->set.postfields ?
+    data->set.postfields : "";
+  const char *content_type = Curl_checkheaders(conn, "Content-Type");
+  unsigned char sha_d[32];
+  char sha_str[65];
+  char *cred_scope = NULL;
+  char *signed_headers = NULL;
+  char request_type[REQUEST_TYPE_L];
+  char *canonical_hdr = NULL;
+  char *canonical_request = NULL;
+  char *str_to_sign = NULL;
+  unsigned char tmp_sign0[32] = {0};
+  unsigned char tmp_sign1[32] = {0};
+  char *auth = NULL;
+  char *tmp;
+  #ifdef DEBUGBUILD
+  char *force_timestamp;
+  #endif
+
+  (void) proxy;
+
+  if(Curl_checkheaders(conn, "Authorization")) {
+    /* Authorization already present, Bailing out */
+    return CURLE_OK;
+  }
+
+  if(!customrequest)
+    customrequest = "POST";
+  if(content_type) {
+    content_type = strchr(content_type, ':');
+    if(!content_type)
+      return CURLE_FAILED_INIT;
+    content_type += 1;
+    /* Skip whitespace now */
+    while(*content_type == ' ' || *content_type == '\t')
+      ++content_type;
+  }
+
+  /* Get Parameter
+     Google and Outscale use the same OSC or GOOG,
+     but Amazon use AWS and AMZ for header arguments */
+  tmp = strchr(provider, ':');
+  if(tmp) {
+    *tmp = 0;
```

Comment:
Doesn't this write a byte in an input string that isn't meant for writing in? It will make it not work in a second invoke, right?